conda-store-server/conda_store_server/server/views/api.py

```diff
@@ -935,6 +935,64 @@ async def api_put_build(
         }
 
 
+@router_api.put(
+    "/build/{build_id}/cancel/",
+    response_model=schema.APIAckResponse,
+)
+async def api_put_build_cancel(
+    build_id: int,
+    request: Request,
+    conda_store=Depends(dependencies.get_conda_store),
+    auth=Depends(dependencies.get_auth),
+):
+    with conda_store.get_db() as db:
+        build = api.get_build(db, build_id)
+        if build is None:
+            raise HTTPException(status_code=404, detail="build id does not exist")
+
+        auth.authorize_request(
+            request,
+            f"{build.environment.namespace.name}/{build.environment.name}",
+            {Permissions.BUILD_CANCEL},
+            require=True,
+        )
+
+        if conda_store.celery_app.control.inspect().ping() is None:
+            raise HTTPException(
+                status_code=409,
+                detail=(
+                    "conda-store celery broker does not support task cancelation. "
+                    "Use redis or rabbitmq message queues. "
+                    "See docs for a more detailed explanation"
+                ),
+            )
+
+        conda_store.celery_app.control.revoke(
+            [
+                f"build-{build_id}-conda-env-export",
+                f"build-{build_id}-conda-pack",
+                f"build-{build_id}-docker",
+                f"build-{build_id}-environment",
+            ],
+            terminate=True,
+            signal="SIGTERM",
+        )
+
+        from conda_store_server.worker import tasks
+
+        tasks.task_cleanup_builds.si(
+            build_ids=[build_id],
+            reason=f"""
+    build {build_id} marked as FAILED due to being canceled from the REST API
+    """,
+        ).apply_async(countdown=5)
```

Comment:
Could maybe use a comment to explain that we have to wait 5 seconds to ensure enough time for the task to actually be canceled.